freebsd: Separate package set contents from construction

Ericson2314 opened the pull request:
## Description of changes

~~This creates multiple package sets for different versions. The other versions don't yet work, but that will be fixed in subsequent PRs.~~

This refactors to create the beginnings of what could become support for multiple package sets / multiple versions of FreeBSD later.

CC @rhelmot @artemist. I am pushing from @rhelmot's repo (thanks for the perms!) so all 3 of us can collaborate.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ericson2314 committed:
generateSplicesForMkScope: Support dot in attribute path

Only the list form supports the full domain of possible attribute names.

Ericson2314 committed:
freebsd: Include more version info and update

We want to know the minor versions.

We want to know the patch versions too, where possible. We get those
either from the branch when it is the form `RELEASE-p<patch>`, or from
the tag when its in the form `release/<major>.<minor>.<patch>`.

Ericson2314 committed:
freebsdBranches{,Cross}: Init

This creates multiple package sets for different versions. The other
versions don't yet work, but that will be fixed in subsequent PRs.

Push versionData into package set so that it can be overridden

Co-Authored-By: Audrey Dutcher <audrey@rhelmot.io>
Co-Authored-By: Artemis Tosini <me@artem.ist>

rhelmot commented:
I've taken a few passes through this and the only thing I can see that's even a little bit objectionable was some dead code, which I removed. Thanks!!

rhelmot commented:
Upon yet further inspection and noticing the ofborg failures I found that the addition of the "patch" version attr was half baked. Not all branches had this attribute, so I filled it in with 0 in the default case. You may wish to fix this in the update.py script instead. Moreover, to distinguish the various branches (release/14.0.0 vs releng/14.0) I added the branch disposition to the end of the version string (14.0.0-RELEASE vs 14.0.0-RELEASE-p6).

alyssais commented:
Can you explain why we want multiple versions?

Ericson2314 commented:
> Upon yet further inspection and noticing the ofborg failures I found that the addition of the "patch" version attr was half baked. Not all branches had this attribute, so I filled it in with 0 in the default case.

Actually, I purposefully wanted not all versions to have it it. The problem was I did the `toString` *after* the non-`null` check.

Ericson2314 commented:
I removed `branch` from the version for now, because I don't want any hashes in the default version to change. But we can add that back later. Also rebased now that the other PR is merged.

Ericson2314 commented:
I also see the `RELEASE-p<patch>` thing @artemist tried to explain before, which I missed. I've updated the script to also extract the patch version from it.

`main` and the `stable/*` branches still lack a `patch` version, however.

Ericson2314 commented:
> Can you explain why we want multiple versions?

@artemist and @rhelmot this one is for you.

I assume it is because you want to both support the last release, and work on the bleeding edge / dogfood things recently upstreamed. I am broadly deferential to that. We have multiple `linuxPackages` for the same reason.

But, all things equal, Nixpkgs does strive to have fewer not more versions of the same package. So there should be a clear answer here if we're going to do this :).

Ericson2314 commented:
(I have one more nix code reworking to do. I want the splicing to take place in `versions.nix` so `package-set.nix` just returns a function and "doesn't know what its own name is.

rhelmot committed:
freebsd/{pkgs,patches}: Move patches to version-dependent patches dir

rhelmot commented:
> Can you explain why we want multiple versions?

Several enterprise appliances (TrueNAS and PFSense come to mind) are based off of relatively shallow forks of FreeBSD. TrueNAS in particular is known for lagging a version behind the latest FreeBSD offering, so I wanted to not leave those users in the dust. The application for configuring a FreeBSD appliance with nix is, of course, being able to build app jails (not a feature we have right now but I've been slowly building up to it).

And, of course, I would like to be able to work on the bleeding edge for other purposes. My vision was that the only things we would 100% make sure worked would be `freebsd14` and `freebsd13` (pending the release of 15 of course), and `freebsd` would be the latest of these two.

I'm noticing suddenly the `patchesRoot` variable is missing. I'm not sure whether that was an explicit rejection or just considered to not be part of this PR - it is in fact essential for being able to support multiple versions, since different versions will need different sets of patches. I've just added it.

I was thinking a little bit about the feasibility of being able to override the entire splice with your own source information and `patchesRoot` so that you could theoretically work off of a fork if you're willing to play debugger for a while, and then hiding both `freebsdGit` and `freebsdBranches` from the consumer of nixpkgs. That might better signal our support goal, though I am genuinely clueless about how it might interact with John's beautiful splices.

Adjacent to this goal is the goal of moving the freebsd ports tree. I think if compatibility for freebsd-ports is at all present in nixpkgs, it should live outside the freebsd package set due to the difficulties in switching its version in tandem with the freebsd-base version.

rhelmot commented:
With the most recent commit I was in fact able to start building freebsd14 code on this branch, with only overriding the freebsd scope with versionData, source, and patchesRoot!

I think this opens the door for the more limited interface I described 30min ago that more accurately describes our support capabilities.

rhelmot commented:
I think the ofborg failure is due to it recursing infinitely into buildFreebsd. You can try fixing this with dontRecursIntoAttrs (?) or by simply removing buildFreebsd from the scope, though that would make it un-overridable. I have to log off for the day so good luck to whoever gets to this first :)

Ericson2314 commented:
> `dontRecursIntoAttrs`

Good call, that is what `pkgs/top-level/stage.nix` does.

Ericson2314 commented:
> That might better signal our support goal, though I am genuinely clueless about how it might interact with John's beautiful splices.

I think it should be fine.

If you are saying `pkgs/os-specific/bsd/freebsd/versions.nix` should define the whitelisted shorthands (like `freebsd13`, `freebsd14`, `freebsdGit`, etc.) that's fine. The only requirement is that `generateSplicesForMkScope` be given the attribute path to where the package set ends up being bound, whatever that may be.

rhelmot commented:
From inspecting the lib-tests code, the only way I could find to squash the recursion error (dontRecurseIntoAttrs was not working) was to add a manual exception. Not sure if this is intended.

rhelmot committed:
freebsd: Only export a limited number of attributes to top-level

rhelmot commented:
The PR title is now inaccurate, but I'm ready to merge if ofborg blesses us. we now only export `freebsd{13,14,}{Cross,}`.

Ericson2314 commented on pkgs/os-specific/bsd/freebsd/versions.nix:
Some of the indentation in here looks a little bit funky?

Ericson2314 replied:
Can also rename this to `default.nix`, signifying it's the "entry point" file for this directory.

Ericson2314 commented on pkgs/os-specific/bsd/freebsd/versions.nix:
```diff
@@ -0,0 +1,56 @@
+      Set one with the NIXPKGS_FREEBSD_BRANCH environment variable or by setting `nixpkgs.config.freebsdBranch`.
+  '';
+  branches = callPackage ./default-branch.nix {};
+  exportedAttrSetsNative = lib.flip lib.mapAttrs branches (_: branch: freebsdBranches.${branch} or (badBranchError branch));
+  exportedAttrSetsCross = lib.flip lib.mapAttrs' branches (
```
Actually the first ones are not necessarily "native". Rather the whole `*Cross` convention is a stupid relic that only exists because the weird bad GCC bootstrapping. Hopefully these new names, while cryptic, are at least accurate. 
```suggestion
  exportedAttrSetsRegular = lib.flip lib.mapAttrs branches (_: branch: freebsdBranches.${branch} or (badBranchError branch));
  exportedAttrSetsPreLibc = lib.flip lib.mapAttrs' branches (
```

Ericson2314 commented:
@alyssais does the explanation of the motivation that @rhelmot gave sound good to you?

alyssais commented:
> We have multiple `linuxPackages` for the same reason.

This isn't totally analogous, because most packages (apart from kernel modules) don't care about linuxPackages.  (They might use linuxHeaders, but we don't provide multiple versions of that because it's backwards-compatible.)  But here, _all_ packages would have to be built separately for each FreeBSD release.

How often is building the "bleeding edge" something normal FreeBSD users are likely to want to do?  Wouldn't it generally be fine to build packages for the oldest version of FreeBSD we want to support, since we've established that FreeBSD is backwards compatible?

We don't provide any special mechanism to allow Linux developers to build against new versions of the kernel headers, but it's simple enough to do in an overlay without needing a global config option (which are generally difficult to work with in Nixpkgs due to us not being able to specify global defaults, among other things):

```nix
final: super: {
  linuxHeaders = linuxHeaders.overrideAttrs ({ ... }: {
    # …
  })
}
```

Could we make FreeBSD similar?

```nix
final: super: {
  freebsd = freebsd.override {
    version = "…";
    hashes = { /* … */ };
  };
}
```

rhelmot commented:
Okay, that's fine. Are you opposed to freebsd13 and freebsd14 to signal our support capabilities?

alyssais commented:
Can we start with just one version in Nixpkgs, and discuss expanding it separately once we have that version working well?

rhelmot committed:
freebsd: Only support branch selection via override... for now.

Remove config element for freebsdBranch

rhelmot commented:
I have performed the requested operation. Again, I don't know if I did it right with respect to splicing, but I'm a little more confident this time!

Ericson2314 commented:
I fixed the formatting issue by running the RFC formatter on the new files.

Ericson2314 commented:
Merged since this is no longer changing public interfaces.

sternenseemann commented:
Fix for eval regression https://github.com/NixOS/nixpkgs/pull/313110
